File: engineering/concurrency/01-mutex-and-lock/skeleton/python/main.py
# 并发模型：threading（真多线程）
import threading

class SafeCounter:
    def __init__(self):
        self._count = 0
        # TODO: 在此添加必要的同步原语

    def inc(self):
        # TODO: 获取锁、递增、释放锁
        self._count += 1

    def value(self):
        # TODO: 获取锁、读取、释放锁
        return self._count

def worker(counter: SafeCounter, m: int):
    for _ in range(m):
        counter.inc()

def main():
    n, m = 100, 1000
    counter = SafeCounter()
    threads = []

    for _ in range(n):
        t = threading.Thread(target=worker, args=(counter, m))
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    expected = n * m
    actual = counter.value()
    if actual == expected:
        print(f"PASS: count = {actual} (expected {expected})")
    else:
        print(f"FAIL: count = {actual} (expected {expected})")

if __name__ == "__main__":
    main()

File: engineering/concurrency/01-mutex-and-lock/skeleton/go/main.go
package main

import (
	"fmt"
	"sync"
)

// SafeCounter 是线程安全的计数器。
// TODO: 在此添加必要的同步原语字段。
type SafeCounter struct {
	// TODO
	count int
}

// Inc 对计数器递增 1。
func (c *SafeCounter) Inc() {
	// TODO: 获取锁、递增、释放锁
	c.count++
}

// Value 返回当前计数值。
func (c *SafeCounter) Value() int {
	// TODO: 获取锁、读取、释放锁
	return c.count
}

func main() {
	const n = 100  // 并发执行单元数量
	const m = 1000 // 每个单元递增次数

	var wg sync.WaitGroup
	counter := &SafeCounter{}

	for i := 0; i < n; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := 0; j < m; j++ {
				counter.Inc()
			}
		}()
	}

	wg.Wait()

	expected := n * m
	actual := counter.Value()
	if actual == expected {
		fmt.Printf("PASS: count = %d (expected %d)\n", actual, expected)
	} else {
		fmt.Printf("FAIL: count = %d (expected %d)\n", actual, expected)
	}
}
